### PR TITLE
[ML] Allow data frame analytics memory estimation on non-ML nodes

### PR DESCRIPTION
**Pull request by droberts195**
The data frame analytics memory estimation process is very small
and short-lived. Therefore, it isn't so bad for it to run on a
node that's not an ML node.

This PR allows the data frame analytics memory estimation process
to run on master, data or ingest nodes in addition to ML nodes. If the
node receiving the _explain request is any of these then it handles
the request, avoiding a transfer within the cluster. Otherwise the
request is transferred to an ML node if there is one. If there is
no ML node then the request is transferred to a master-eligible node.

Replacement for #67840

### CODE REVIEW

File: x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle [hunk 1]
```diff
@@ -52,6 +52,10 @@ testClusters.all {
   setting 'xpack.ml.min_disk_space_off_heap', '200mb'
   setting 'indices.lifecycle.history_index_enabled', 'false'
   setting 'slm.history_index_enabled', 'false'
+  // To spice things up a bit, one of the nodes is not an ML node
+  nodes.'javaRestTest-0'.setting 'node.roles', '["master","data","ingest"]'
+  nodes.'javaRestTest-1'.setting 'node.roles', '["master","data","ingest","ml"]'
+  nodes.'javaRestTest-2'.setting 'node.roles', '["master","data","ingest","ml"]'
```
Comment (droberts195) [inline x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle:58]: We can't use an internal cluster test to test this functionality, because internal cluster tests cannot run native processes.  Instead, making one of the nodes in the multi-node cluster a non-ML node means one third of the `_explain` requests made by the test harness will go to it, and this ensures we have some test coverage for this change.

File: x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle [hunk 2]
```diff
@@ -52,6 +52,10 @@ testClusters.all {
   setting 'xpack.ml.min_disk_space_off_heap', '200mb'
   setting 'indices.lifecycle.history_index_enabled', 'false'
   setting 'slm.history_index_enabled', 'false'
+  // To spice things up a bit, one of the nodes is not an ML node
```
Comment (benwtrent) [inline x-pack/plugin/ml/qa/native-multi-node-tests/build.gradle:55]: 🌶️

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java [hunk 3]
```diff
@@ -206,31 +206,34 @@ private void estimateMemoryUsage(Task task,
...
         for (DiscoveryNode node : clusterState.getNodes()) {
             if (MachineLearning.isMlNode(node)) {
                 return Optional.of(node);
             }
         }
+        for (DiscoveryNode node : clusterState.getNodes()) {
+            if (node.isDataNode() || node.isIngestNode()) {
+                return Optional.of(node);
+            }
+        }
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java:236]: Do we need to worry about the ml plugin being enabled or not?

Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java:236]: Yes, good point.  We demand that the ML plugin is enabled on master-eligible nodes, so I think the only safe option will be to redirect to one of these.  I will try to get it to prefer the master-eligible nodes that aren't the active master.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java [hunk 4]
```diff
@@ -206,31 +206,34 @@ private void estimateMemoryUsage(Task task,
     }
 
     /**
-     * Finds the first available ML node in the cluster and redirects the request to this node.
+     * Finds the a suitable node in the cluster we can run on, and redirects the request to this node.
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java:209]: nit: remove "the" after "Finds"

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java (hunk 3 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java]: OK, I think that will work. 

It would be cool if we could determine if the plugin is enabled or not given the info in the discovery node data. That way we are not restricted to only master nodes.

But, I am not sure that is possible. 

So, on a minimal deployment (just one node), this will work as we always need at least one master node :).

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java (hunk 3 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java]: The only times a redirection is needed now are:

1. When the connection comes from a transport client (impossible in 8.x), because this code will run on the transport client node
2. When the connection comes from a node client (only possible in tests now, but a crucial case for getting CI to pass), because this code will run on the node client node
3. When the connection comes from the REST API but goes to a pure coordinating node in the cluster - we could avoid this case too in 8.x if there was an easy way to distinguish node client nodes from coordinating nodes

There is talk of maybe adding a dedicated `coordinating` node role in the future.  That would satisfy the requirement of 3.

The TL;DR is that redirections will be rarer in production than in our tests, and can probably be made even rarer in the future.  So I don't think it's a good idea to add even more attributes to the cluster state just for this.
